### ISSUE TITLE
[Bug]: No parallelization across multiple-gpus at inference (DDP)

### ISSUE THREAD

**Issue by matthewkperez** [speechbrain/speechbrain#2014]
### Describe the bug

When running the recipe found in `recipes/LibriSpeech/ASR/transformer` I notice that during evaluation the test set is not parallelized over the multiple gpus such as the case during training. Instead at inference, each GPU runs through the entire test set, which seems slow and inefficient.

Is this the intended behavior?

### Expected behaviour

Test set is split across multiple gpus at test time

### To Reproduce

```
CUDA_VISIBLE_DEVICES=0,1,2 python -m torch.distributed.launch --nproc_per_node=3 train.py hparams/transformer.yaml --distributed_launch --distributed_backend=nccl --find_unused_parameters
```

Note: the only change I made to the recipe yaml was turning dynamic batching off.

**Comment by matthewkperez** [comment]
I see that the train_data_loader has its own function _train_loader_specifics(). But why is this function specific to the training set only? Why not have this the test and validation sets also leverage multiple gpus using DistributedSampler()?

**Comment by Gastron** [comment]
This is intended behaviour: The issue is in inference is the collection of metrics from all nodes. This is difficult to accomplish as a default behaviour that would work for _every_ case while maintaining high flexibility. At the same time, inference often runs fast anyway, so it often doesn't matter. 

However, this can relatively easily be done for any given metric, so any recipe could implement this to save time in inference.

**Comment by matthewkperez** [comment]
Found a solution. Posting below in case this can be helpful to others.

So for my purposes, I have a a single machine with multiple gpus that I'd like to leverage for decoding. The high-level approach is to run inference on across these gpus and then have a separate summarize() for each metric of interest that uses the torch.distributed.reduce() to collect data across the distributed processes.

Below is an example for AccuracyStats()
```
def summarize_dist(self, device):
    """Computes the accuracy metric."""
    # local values
    correct_tensor = torch.tensor([self.correct]).to(device)
    total_tensor = torch.tensor([self.total]).to(device)

    # reduce tensors across all processes
    dist.reduce(correct_tensor, dst=0, op=dist.ReduceOp.SUM, group=dist.group.WORLD)
    dist.reduce(total_tensor, dst=0, op=dist.ReduceOp.SUM, group=dist.group.WORLD)

    # The accuracy is only computed on the master process
    if dist.get_rank() == 0:
        accuracy = correct_tensor / total_tensor
        return accuracy.item()

    return None
```

For the main ASR_brain class, use this the summarize_dist call in on_stage_end()
```
def on_stage_end(self, stage, stage_loss, epoch):
        """Gets called at the end of a epoch."""
        # Compute/store important stats
        stage_stats = {"loss": stage_loss}
        if stage == sb.Stage.TRAIN:
            self.train_stats = stage_stats
        else:
            stage_stats["ACC"] = self.acc_metric.summarize_dist(self.device)

...... etc
```

Additionally make sure that your test_dataset is passed through train_loader_specifics()
```
def make_dataloader(
        self, dataset, stage, ckpt_prefix="dataloader-", **loader_kwargs
    ):
        loader_kwargs = self._train_loader_specifics(dataset, loader_kwargs)

..... etc
```
